### ISSUE TITLE
Additional overloads in some System.Math methods to support more types

### ISSUE THREAD

**Issue by varocarbas** [dotnet/coreclr#3204]
`Math.Round` can only deal with `decimal` and `double` arguments, what is quite arbitrary as far as these functionalities might also be used with integer types (e.g., `int` or `long`). There is no difference between removing all the digits after 6 in 1.5658 and in 15658, but .NET treats both situations differently. In the first case, it is possible to use `Math.Round` right away; in the second case, two additional steps would be required: either converting to a decimal type or to string and use `Substring`; and then back to the original integer type.

Other function with restricted input types is `Math.Pow` which only accepts `double` type inputs. I have been in quite a few scenarios where I had to convert `decimal` variables to `double` (and then back to `decimal`) just to use this function. Equivalent ideas are applicable to other `System.Math` methods which only work with `double` variables.

CLARIFICATION: `Math.Truncate` doesn't deliver the standard (i.e., mathematical) truncation (https://en.wikipedia.org/wiki/Truncation), which is done by `Math.Floor`. What this function does is removing all the digits after the decimal point and that's why its current only-supporting-`decimal`-&-`double` version is fine.

CLARIFICATION 2: `Math.Floor` doesn't deliver truncation either. The truncated value for 1.23456 to 3 decimal positions is 1.234. But `Math.Floor` delivers 1 in this case (i.e., truncation to the closest integer). In .NET, the closest equivalence to mathematical truncation is delivered by the `Math.Round` overloads taking `MidpointRounding` arguments; but even those might not deliver a proper truncation. 

BOTTOM LINE: the aforementioned clarifications shouldn't represent a big deal here as far as programming languages are meant to be practical tools; and the current implementation is practical enough (except for the aforementioned limitations in the supported types). At least, this is the case for me: I have use `System.Math` functions a lot, without having ever thought about all these issues (= never found a problem). On the other hand, I might pull a function delivering mathematical truncation (i.e., just removing all the last numbers after certain point, regardless of the decimal separator and without rounding).

**Comment by varocarbas** [comment]
I took this question as an excuse to know more about the native part (i.e., all the referred `System.Math` methods) in CoreCLR. After struggling for a while with the code, I wasn't able to find anything, other than a complicated system of references among a huge amount of files.

But then I saw this Hans Passant answer in StackOverflow: http://stackoverflow.com/questions/8870442/how-is-math-pow-implemented-in-net-framework

I guess that he is right (this is usually the case with him), what explains the no-code issue (+ why not having included more overloads before). I mean all what is near the "pow() though, Microsoft purchased that code from Intel" bit; ideas which are also likely to be applicable to other `System.Math` methods.

Under these conditions, I think that removing C++ from the picture and implementing the new overloads in C# is the best alternative. Perhaps it might be slightly slower (not sure because I don't even know how the original C++ code looks like), but this will not be an issue because of not having other alternative. For example: implementing the new `Math.Pow(decimal x, decimal y)` in C# and letting anyone not liking it continue using the old `Math.Pow(double x, double y)`, together with the type conversion (well... the new approach will certainly be faster).

I am much more experienced in C# than in C++ and I would like to take care of this implementation (in case of going ahead).

**Comment by weshaggard** [comment]
Please follow the process outlined at http://aka.ms/apireview and file the API proposal in the http://github.com/dotnet/corefx repo.

**Comment by varocarbas** [comment]
@weshaggard OK. I will do it right now. Thanks for the suggestion.

**Comment by sergiy-k** [comment]
Marking as Future to match https://github.com/dotnet/corefx/issues/6308.

**Comment by varocarbas** [comment]
This proposal has been discussed, analysed, rejected and closed in other two CoreFX issues ([last and most relevant one](https://github.com/dotnet/corefx/issues/13834)). That's why keeping this issue open makes no sense.
